<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
</head>

<body>

</body>

</html>
<script src="http://libs.baidu.com/jquery/2.0.0/jquery.min.js"></script>
<script>
    let p = new Promise(function (res, rej) { //promise 状态一经改变就不在进行任何改变。
        // setTimeout(() => {
        //     console.log(qqq);

        // }, 2000)
        // $.ajax({
        //     url: './data.json',
        //     success: function (data) {
        //         res(data)
        //     },
        //     error: function (err) {
        //         rej(err)
        //     }
        // })
        res(qqq)
    }).then(function () { //连续的then 第二个then的形参是第一个then的return 返回值
        console.log(1);

    }, function (error) {
        console.log(0);

    }).catch(function (eer) {
        console.log(6666);//then写两个的时候，错误信息被then的第二个函数捕获，catch就不会捕获错误信息了

    }).finally(() => {
        console.log(999);  //finally 不管成功还是失败， 都要执行的代码 我们可以放到这个函数中

    })
    //then catch  finally          这三个方法都是在promise的原型上  都是异步函数
    //all rece         resolve reject




    var p1 = new Promise(function (res, rej) {
        setTimeout(() => {
            res(111);
        }, 2000);
    })
    var p2 = new Promise(function (res, rej) {
        setTimeout(() => {
            res(222);
        }, 3000);
    })
    var p3 = new Promise(function (res, rej) {
        setTimeout(() => {
            rej(333);
        }, 4000);
    })
    //所有的都成功 才会走成功的回调函数    只要有一个失败就直接走对应的失败函数
    //exector 函数是一个同步执行的函数  then catch 里面的函数是异步执行的
    Promise.all([p1, p2, p3]).then((data) => {
        console.log("data", data);

    }, (err) => {
        console.log("err", err);

    });

    Promise.race([p1,p2,p3]).then((data)=>{
        console.log('success',data);
        
    },(err)=>{
        console.log('err',err);
        
    })
    Promise.allSettled([p1, p2, p3]).then((data) => {
        console.log(data)
    });
    Promise.resolve(3333).then((data) => {
        console.log(data);

    });

</script>